File: binding/src/libqtemporaryfile.cpp
#include <QFile>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QTemporaryFile>
#include <qtemporaryfile.h>
#include "libqtemporaryfile.h"
#include "libqtemporaryfile.hxx"

QTemporaryFile* QTemporaryFile_new() {
	 return new VirtualQTemporaryFile();
}

QTemporaryFile* QTemporaryFile_new2(const libqt_string templateName) {
	 return new VirtualQTemporaryFile(QString::fromUtf8(templateName.data, templateName.len));
}

QTemporaryFile* QTemporaryFile_new3(QObject* parent) {
	 return new VirtualQTemporaryFile(parent);
}

QTemporaryFile* QTemporaryFile_new4(const libqt_string templateName, QObject* parent) {
	 return new VirtualQTemporaryFile(QString::fromUtf8(templateName.data, templateName.len), parent);
}

libqt_string QTemporaryFile_Tr(const char* s) {
	QString _ret = QTemporaryFile::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QTemporaryFile_AutoRemove(const QTemporaryFile* self) {
	return self->autoRemove();
}

void QTemporaryFile_SetAutoRemove(QTemporaryFile* self, bool b) {
	self->setAutoRemove(b);
}

bool QTemporaryFile_Open(QTemporaryFile* self) {
	return self->open();
}

libqt_string QTemporaryFile_FileName(const QTemporaryFile* self) {
	QString _ret = self->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QTemporaryFile_FileTemplate(const QTemporaryFile* self) {
	QString _ret = self->fileTemplate();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QTemporaryFile_SetFileTemplate(QTemporaryFile* self, const libqt_string name) {
	self->setFileTemplate(QString::fromUtf8(name.data, name.len));
}

bool QTemporaryFile_Rename(QTemporaryFile* self, const libqt_string newName) {
	return self->rename(QString::fromUtf8(newName.data, newName.len));
}

QTemporaryFile* QTemporaryFile_CreateNativeFile(const libqt_string fileName) {
	return QTemporaryFile::createNativeFile(QString::fromUtf8(fileName.data, fileName.len));
}

QTemporaryFile* QTemporaryFile_CreateNativeFile2(QFile* file) {
	return QTemporaryFile::createNativeFile(*file);
}

libqt_string QTemporaryFile_Tr2(const char* s, const char* c) {
	QString _ret = QTemporaryFile::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QTemporaryFile_Tr3(const char* s, const char* c, int n) {
	QString _ret = QTemporaryFile::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
libqt_string QTemporaryFile_QBaseFileName(const QTemporaryFile* self) {
	auto* vqtemporaryfile = dynamic_cast<const VirtualQTemporaryFile*>(self);
	if (vqtemporaryfile && vqtemporaryfile->isVirtualQTemporaryFile) {
vqtemporaryfile->setQTemporaryFile_FileName_IsBase(true);
	QString _ret = vqtemporaryfile->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QTemporaryFile_OnFileName(const QTemporaryFile* self, intptr_t slot) {
	auto* vqtemporaryfile = dynamic_cast<const VirtualQTemporaryFile*>(self);
	if (vqtemporaryfile && vqtemporaryfile->isVirtualQTemporaryFile) {
vqtemporaryfile->setQTemporaryFile_FileName_Callback(reinterpret_cast<VirtualQTemporaryFile::QTemporaryFile_FileName_Callback>(slot));
}
}

// Derived class handler implementation
bool QTemporaryFile_Open2(QTemporaryFile* self, int flags) {
	auto* vqtemporaryfile = dynamic_cast<VirtualQTemporaryFile*>(self);
	if (vqtemporaryfile && vqtemporaryfile->isVirtualQTemporaryFile) {
	return vqtemporaryfile->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
	} else {
	return self->QTemporaryFile::open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
}
}

// Base class handler implementation
bool QTemporaryFile_QBaseOpen2(QTemporaryFile* self, int flags) {
	auto* vqtemporaryfile = dynamic_cast<VirtualQTemporaryFile*>(self);
	if (vqtemporaryfile && vqtemporaryfile->isVirtualQTemporaryFile) {
vqtemporaryfile->setQTemporaryFile_Open2_IsBase(true);
	return vqtemporaryfile->open(static_cast<QFlags<QIODeviceBase::OpenModeFlag>>(flags));
}
}

// Auxiliary method to allow providing re-implementation
void QTemporaryFile_OnOpen2(QTemporaryFile* self, intptr_t slot) {
	auto* vqtemporaryfile = dynamic_cast<VirtualQTemporaryFile*>(self);
	if (vqtemporaryfile && vqtemporaryfile->isVirtualQTemporaryFile) {
vqtemporaryfile->setQTemporaryFile_Open2_Callback(reinterpret_cast<VirtualQTemporaryFile::QTemporaryFile_Open2_Callback>(slot));
}
}

void QTemporaryFile_Delete(QTemporaryFile* self) {
    delete self;
}

File: binding/include/libqtemporaryfile.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qtemporaryfile.html
pub const qtemporaryfile = struct {

    /// New constructs a new QTemporaryFile object.
    pub fn New() QtC.QTemporaryFile {
        return qtc.QTemporaryFile_new();
    }


    /// New2 constructs a new QTemporaryFile object.
    pub fn New2(templateName: []const u8) QtC.QTemporaryFile {
        const templateName_str = qtc.libqt_string{
    .len = templateName.len,
    .data = templateName.ptr,
};
return qtc.QTemporaryFile_new2(templateName_str);
    }


    /// New3 constructs a new QTemporaryFile object.
    pub fn New3(parent: ?*anyopaque) QtC.QTemporaryFile {
        return qtc.QTemporaryFile_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QTemporaryFile object.
    pub fn New4(templateName: []const u8, parent: ?*anyopaque) QtC.QTemporaryFile {
        const templateName_str = qtc.libqt_string{
    .len = templateName.len,
    .data = templateName.ptr,
};
return qtc.QTemporaryFile_new4(templateName_str, @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryFile_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporaryfile.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AutoRemove(self: ?*anyopaque, ) bool {
        return qtc.QTemporaryFile_AutoRemove(@ptrCast(self));
    }

    pub fn SetAutoRemove(self: ?*anyopaque, b: bool) void {
        qtc.QTemporaryFile_SetAutoRemove(@ptrCast(self), b);
    }

    pub fn Open(self: ?*anyopaque, ) bool {
        return qtc.QTemporaryFile_Open(@ptrCast(self));
    }

    pub fn FileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryFile_FileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporaryfile.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnFileName(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QTemporaryFile_OnFileName(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryFile_QBaseFileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporaryfile.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn FileTemplate(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryFile_FileTemplate(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporaryfile.FileTemplate: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetFileTemplate(self: ?*anyopaque, name: []const u8) void {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
qtc.QTemporaryFile_SetFileTemplate(@ptrCast(self), name_str);
    }

    pub fn Rename(self: ?*anyopaque, newName: []const u8) bool {
        const newName_str = qtc.libqt_string{
    .len = newName.len,
    .data = newName.ptr,
};
return qtc.QTemporaryFile_Rename(@ptrCast(self), newName_str);
    }

    pub fn CreateNativeFile(fileName: []const u8) QtC.QTemporaryFile {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QTemporaryFile_CreateNativeFile(fileName_str);
    }

    pub fn CreateNativeFile2(file: ?*anyopaque) QtC.QTemporaryFile {
        return qtc.QTemporaryFile_CreateNativeFile2(@ptrCast(file));
    }

    pub fn Open2(self: ?*anyopaque, flags: i32) bool {
        return qtc.QTemporaryFile_Open2(@ptrCast(self), @intCast(flags));
    }

    /// Allows for overriding the related default method
    pub fn OnOpen2(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) bool) void {
        qtc.QTemporaryFile_OnOpen2(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseOpen2(self: ?*anyopaque, flags: i32) bool {
        return qtc.QTemporaryFile_QBaseOpen2(@ptrCast(self), @intCast(flags));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryFile_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporaryfile.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTemporaryFile_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtemporaryfile.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QTemporaryFile_Delete(@ptrCast(self));
    }
};
